src/Stratis.Bitcoin/Consensus/ConsensusManger.cs

```diff
@@ -74,13 +117,388 @@ public class ConsensusManager
                 utxoHash = await this.consensusRules.RewindAsync().ConfigureAwait(false);
             }
 
-            this.concurrentChain.SetTip(this.Tip);
+            lock (this.treeLock)
+            {
+                this.chainedHeaderTree.Initialize(this.Tip, !blockStoreDisabled);
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// A list of headers are presented from a given peer,
+        /// we'll attempt to connect the headers to the tree and if new headers are found they will be queued for download.
+        /// </summary>
+        /// <param name="peerId">The peer that is providing the headers.</param>
+        /// <param name="headers">The list of new headers.</param>
+        /// <returns>The last chained header that is connected to the tree.</returns>
+        public ChainedHeader HeadersPresented(int peerId, List<BlockHeader> headers)
+        {
+            this.logger.LogTrace("({0}:{1},{2}.{3}:{4})", nameof(peerId), peerId, nameof(headers), nameof(headers.Count), headers.Count);
+
+            ConnectNewHeadersResult newHeaders = null;
+
+            lock (this.treeLock)
+            {
+                newHeaders = this.chainedHeaderTree.ConnectNewHeaders(peerId, headers);
+                this.blockPuller.NewTipClaimed(peerId, newHeaders.Consumed); // TODO: should this be inside the lock?
```

Comment:
Yes, it is just the name of the lock that is slightly misleading. We need to prevent race conditions with peer disconnections.

Reply:
ok so basically puller locks are also in CM thats fine, we just need to make sure `NewTipClaimed` is a small and fast method.

Reply:
Not really, block puller has its own locks but here we need to synchronise these two components together